### PR TITLE
refactor network generate

### PR DESCRIPTION
**Pull request by Junnplus**
Signed-off-by: ye.sijun <junnplus@gmail.com>

Fixes: https://github.com/containerd/nerdctl/issues/671

### CODE REVIEW

File: pkg/netutil/cni_plugin.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	AddrSpace     string      `json:"addressSpace,omitempty"`
+	Subnet        string      `json:"subnet,omitempty"`
+	Address       string      `json:"ipAddress,omitempty"`
+	QueryInterval string      `json:"queryInterval,omitempty"`
+	Routes        []IPAMRoute `json:"routes,omitempty"`
+}
+
+func newWindowsIPAMConfig() *windowsIpamConfig {
+	return &windowsIpamConfig{}
+}
```
Comment (AkihiroSuda) [inline pkg/netutil/cni_plugin.go:171]: Can we move windows-specific stuffs to `*_windows.go`, and linux-specific stuffs to `*_linux.go`?

File: pkg/netutil/netutil_unix.go [hunk 2]
```diff
@@ -23,48 +23,6 @@ const (
 	DefaultNetworkName = "bridge"
 	DefaultID          = 0
 	DefaultCIDR        = "10.4.0.0/24"
+	DefaultCNIDriver   = "bridge"
```
Comment (AkihiroSuda) [inline pkg/netutil/netutil_unix.go:26]: Perhaps s/Driver/Plugin/g

File: pkg/netutil/cni_plugin.go (hunk 1 above)
Comment (Junnplus) [reply pkg/netutil/cni_plugin.go:171]: Here is just some structural definitions, separation will make tricky for `GenerateCNIPlugins`.

File: pkg/netutil/netutil.go [hunk 3]
```diff
@@ -211,3 +214,88 @@ func NerdctlLabels(b []byte) *map[string]string {
...
+func GenerateIPAM(driver string, subnetStr string) (map[string]interface{}, error) {
+	if driver == "" {
+		driver = DefaultIPAMDriver
+	}
+	subnetIP, subnet, err := net.ParseCIDR(subnetStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse subnet %q", subnet)
+	}
+	if !subnet.IP.Equal(subnetIP) {
+		return nil, fmt.Errorf("unexpected subnet %q, maybe you meant %q?", subnet, subnet.String())
```
Comment (AkihiroSuda) [inline pkg/netutil/netutil.go:254]:
```suggestion
		return nil, fmt.Errorf("unexpected subnet %q, maybe you meant %q?", subnetIP, subnet.String())
```

File: pkg/netutil/netutil.go [hunk 4]
```diff
@@ -48,66 +48,69 @@ type CNIEnv struct {
...
 }
 
+const ConfigListTemplate = `{
+  "cniVersion": "0.4.0",
+  "name": "{{.Name}}",
+  "nerdctlID": {{.ID}},
+  "nerdctlLabels": {{.Labels}},
+  "plugins": {{.Plugins}}
+}`
+
```
Comment (AkihiroSuda) [inline pkg/netutil/netutil.go:63]: Can we just marshal a struct into JSON, without using text/template?

Comment (Junnplus) [reply pkg/netutil/netutil.go:63]: SGTM, If you are not in a hurry, I may write another PR to implement.

File: pkg/netutil/netutil_windows.go [hunk 5]
```diff
@@ -20,39 +20,6 @@ const (
 	DefaultNetworkName = "nat"
 	DefaultID          = 0
 	DefaultCIDR        = "10.4.0.0/24"
+	DefaultCNIPlugin   = "nat"
+	DefaultIPAMDriver  = "windows"
```
Comment (fahedouch) [inline pkg/netutil/netutil_windows.go:24]: this disturbs me a bit. Is there really a driver called `windows`?

File: pkg/netutil/netutil.go [hunk 6]
```diff
@@ -211,3 +214,88 @@ func NerdctlLabels(b []byte) *map[string]string {
...
+		ipamConfig = ipamConf
+	case "windows":
+		ipamConf := newWindowsIPAMConfig()
+		ipamConf.Subnet = subnet.String()
+		ipamConf.Routes = append(ipamConf.Routes, IPAMRoute{Gateway: gateway.String()})
+		ipamConfig = ipamConf
+	default:
+		return nil, fmt.Errorf("unsupported ipam driver %q", driver)
+	}
+
```
Comment (fahedouch) [inline pkg/netutil/netutil.go:282]: I think that cases are not balanced. `Windows` is an os and not a driver.

you can make `netutil_windows.go` and `netutil_linux.go`

File: pkg/netutil/netutil.go [hunk 7]
```diff
@@ -48,66 +48,69 @@ type CNIEnv struct {
 }
 
 func DefaultConfigList(e *CNIEnv) (*NetworkConfigList, error) {
-	return GenerateConfigList(e, nil, DefaultID, DefaultNetworkName, DefaultCIDR)
+	ipam, _ := GenerateIPAM("", DefaultCIDR)
+	plugins, _ := GenerateCNIPlugins("", DefaultID, ipam)
```
[the remark this answers is not in the document]
Comment (Junnplus) [reply pkg/netutil/netutil.go:52]: For the default network we can guarantee that it will not error.

File: pkg/netutil/netutil.go (hunk 4 above)
Comment (Junnplus) [reply pkg/netutil/netutil.go]: I'm back, rewrite it with marshal.

File: docs/cni.md [hunk 8]
```diff
@@ -13,6 +13,48 @@ system, the supported CNI plugin types are `nat` only.
...
+      "capabilities": {
+        "portMappings": true
+      }
+    },
+    {
+      "type": "firewall"
+    },
+    {
+      "type": "tuning"
+    }
```
Comment (AkihiroSuda) [inline docs/cni.md:53]: And `isolation` (when installed)

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**